### DISCUSSION TITLE
`os.fork()` causes `jax` to freeze

### DISCUSSION THREAD

**Discussion by homerjed** [jax-ml/jax#21073] [Q&A]
I run a typical `jax` script and this warning pops up

```python
.../lib/python3.12/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
```
after running `import jax; jax.print_environment_info()` I get

```python
jax:    0.4.26
jaxlib: 0.4.26
numpy:  1.26.4
python: 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='kng-cl-nv01', release='5.14.0-362.24.2.el9_3.x86_64', version='#1 SMP PREEMPT_DYNAMIC Sat Mar 30 14:11:54 EDT 2024', machine='x86_64')


$ nvidia-smi
Sat May  4 13:38:20 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A40                     Off |   00000000:61:00.0 Off |                    0 |
|  0%   38C    P0             61W /  300W |     270MiB /  46068MiB |      2%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A   3810861      C   python                                        262MiB |
+-----------------------------------------------------------------------------------------+
```

This was after a fresh reinstall in a new virtual environment. Has anyone got any pointers here? I assume this works for other people and I'm doing something wrong!

**Comment by jakevdp** [comment]
Hi - can you say more about what your "typical jax script" includes? This warning (and the deadlock that it warns about) will occur if your script attempts to use JAX together with multiprocessing. JAX's computational model is not compatible with multiprocessing.

Is it possible your script is attempting to use JAX with some library like `joblib` that invokes `os.fork` in the background?

**Comment by jacksonloper** [comment]
One example of how this comes up is if you run a command in a jupyter notebook, using !, e.g.

!ls -alh veryintriguingpath

Or if you use imageio to write an mpeg file.  Many packages seem to fork as a matter of course when calling subprocess.  It would be nice if Jax didn't rebel when you need to invoke such packages.  But tbh how it all works is honestly quite beyond me, so I'm not sure how/if this can be done.

**Comment by hawkinsp** [comment]
Well, it's a warning. You are free to ignore it and things might work if you get lucky. It is unfortunately correct in this case: calling `os.fork()` from Python is always unsafe. Everything between `fork()` and `exec()` needs to be async signal safe C code, and that's simply never true when you use the Python-level `os.fork()`.

The warning issued on a jupyter `!...` command is actually correct: there is a chance of that deadlocking. It appears `jupyter` uses `ptyprocess` to do that, and there's a long-standing bug in `ptyprocess` about that: https://github.com/pexpect/ptyprocess/issues/43

The right way to fix that problem would be upstream in `ptyprocess`, and the only real fix is to turn that code into a C extension instead of implementing it in Python.

Hope that helps!
